### DISCUSSION TITLE
Add init property to the return type of `atomWithReset()`

### DISCUSSION THREAD

**Discussion by jaesoekjjang** [pmndrs/jotai#2303] [Ideas]
Hello,

In my opinion, if the init property is inferred in the return type of `atomWithReset()`, it would be more convinient in TypeScript.
I would like to change it myself as below, if it was not intended to conceal it.

``` typescript
function atomWithReset<Value>(
  initialValue: Value,
): WritableAtom<Value, SetStateAction<Value> | typeof RESET> & WithInitialValue<Value>
```

Thank you

**Comment by dai-shi** [comment]
Yeah, adding the type for init property is possible. Our of curiosity, what is your use case?

**Reply by jaesoekjjang** [reply]
I used it to make a util function to reset multiple atoms for a formData at once

**Reply by dai-shi** [reply]
I see. Can you share an example of such a util function?
It would be nice if you can use http://www.typescriptlang.org/play to show the TS error.

**Reply by jaesoekjjang** [reply]
Cannot be same since it's company's code, but in the context where the 'init' property was required from outer function, it resembles to this example 

[usePartialReset](https://codesandbox.io/p/sandbox/jotai-usepartialreset-8ykr4n?file=%2Fsrc%2Fhooks%2FusePartialResetAtom.ts%3A4%2C1&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clq9leehz00063b83mjm2op1t%2522%252C%2522sizes%2522%253A%255B88.09775429326288%252C11.90224570673712%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clq9leehz00023b83z211b2rt%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clq9leehz00033b83hebipwrf%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clq9leehz00053b83cp4yq05j%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clq9leehz00023b83z211b2rt%2522%253A%257B%2522id%2522%253A%2522clq9leehz00023b83z211b2rt%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clq9lg38r00023b83yug761da%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A2%252C%2522startColumn%2522%253A19%252C%2522endLineNumber%2522%253A2%252C%2522endColumn%2522%253A19%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clq9lipwo00023b83xmde4p5k%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A4%252C%2522startColumn%2522%253A1%252C%2522endLineNumber%2522%253A4%252C%2522endColumn%2522%253A1%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252Fhooks%252FusePartialResetAtom.ts%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clq9lipwo00023b83xmde4p5k%2522%257D%252C%2522clq9leehz00053b83cp4yq05j%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clq9leehz00043b83furlag5k%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clq9leehz00053b83cp4yq05j%2522%252C%2522activeTabId%2522%253A%2522clq9leehz00043b83furlag5k%2522%257D%252C%2522clq9leehz00033b83hebipwrf%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clq9leehz00033b83hebipwrf%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

**Reply by dai-shi** [reply]
Thanks for sharing. From the code, what I can say is, it doesn't seem very portable to rely on `.init` in the userland code. `.init` is an implementation detail and not considered as public api. None of jotai/utils depend on it except for some that are noted as "HACK".

**Reply by jaesoekjjang** [reply]
Ah, I see. I was misusing it.
Is it possible to share a code you consider the best practice to acheive the same thing what i've tried?

**Reply by dai-shi** [reply]
I wouldn't use atomWithReset. And, I'd prefer atom utils to hooks.

https://tsplay.dev/w8gx4W

```ts
import { atom } from "jotai";

export function atomWithPartialReset<Value extends Record<string, unknown>>(
  initialValue: Value,
) {
  const baseAtom = atom(initialValue);
  return atom(
    (get) => get(baseAtom),
    (get, set, keys: (keyof Value)[]) => {
      const value = { ...get(baseAtom) };
      keys.forEach((key) => {
        value[key] = initialValue[key];
      });
      set(baseAtom, value);
    },
  );
}
```

**Reply by jaesoekjjang** [reply]
Thanks a lot. It really helped me!
